nixos/modules/services/networking/vpp.nix

```diff
@@ -0,0 +1,658 @@
+            hqos-thread ${toString dev'.value.hqos.hqos-thread}
+          ''}
+          ${optionalString (hasValue dev'.value.hqos) "}"}
+        ''}
+        ${optionalString (hasValue dev'.value) "}"}
+      '') (mapAttrsToList (name: value: { name = name; value = value; }) dev)}
+    '' + optionalString (vdev != null) ''
+      ${concatMapStrings (vdev': ''
+        vdev ${vdev'}
+      '') vdev}
+    '' + optionalString (num-mbufs != null) ''
+      num-mbufs ${toString num-mbufs}
+    '' + optionalString no-pci ''
+      no-pci
+    '' + optionalString no-hugetlb ''
+      no-hugetlb
+    '' + optionalString (kni != null) ''
+      kni ${toString kni}
+    '' + optionalString (uio-driver != null) ''
+      uio-driver ${uio-driver}
+    '' + optionalString (socket-mem != null) ''
+      socket-mem ${toString socket-mem}
+    '' + optionalString enable-tcp-udp-checksum ''
+      enable-tcp-udp-checksum
+    '' + optionalString no-multi-seg ''
+      no-multi-seg
+    '' + optionalString no-tx-checksum-offload ''
+      no-tx-checksum-offload
+    '' + optionalString decimal-interface-names ''
+      decimal-interface-names
+    '' + optionalString (log-level != null) ''
+      log-level ${log-level}
+    '';
+  dpdkSection = ''
+    dpdk {
+    ${dpdkSectionConf}
+    }
+  '';
+  cpuSectionConf = with cfg.cpu;
+    optionalString (workers != null) ''
+      workers ${toString workers}
+    '' + optionalString (io != null) ''
+      io ${toString io}
+    '' + optionalString main-thread-io ''
+      main-thread-io
+    '' + optionalString (skip-cores != null) ''
+      skip-cores ${toString skip-cores}
+    '' + optionalString (main-core != null) ''
+      main-core ${toString main-core}
+    '' + optionalString (coremask-workers != null) ''
+      coremask-workers ${coremask-workers}
+    '' + optionalString (corelist-workers != null) ''
+      corelist-workers ${corelist-workers}
+    '' + optionalString (corelist-io != null) ''
+      corelist-io ${corelist-io}
+    '' + optionalString (coremask-hqos-threads != null) ''
+      coremask-hqos-threads ${coremask-hqos-threads}
+    '' + optionalString (corelist-hqos-threads != null) ''
+      corelist-hqos-threads ${corelist-hqos-threads}
+    '' + optionalString (thread-prefix != null) ''
+      thread-prefix ${thread-prefix}
+    '' + optionalString (scheduler-priority != null) ''
+      scheduler-priority ${toString scheduler-priority}
+    '' + optionalString (extraConfig != null) ''
+      ${extraConfig}
+    '';
+  cpuSection = ''
+    cpu {
+    ${cpuSectionConf}
+    }
+  '';
+  startupConf = pkgs.writeText "startup.conf" (removeEmptyLines (''
+    ${unixSection}
+  '' + optionalString (dpdkSectionConf != "") ''
+    ${dpdkSection}
+  '' + optionalString (cpuSectionConf != "") ''
+    ${cpuSection}
+  '' + optionalString (cfg.extraConfig != "") ''
+    ${cfg.extraConfig}
+  ''));
+in
+{
+
+  ###### interface
+
+  options = {
```

Comment:
I would prefer to use [RFC 0042](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md)-style structured settings option, instead of mirroring each config option of VPP inside the NixOS module. This is because a lot of individual NixOS option impact evaluation performance, and also have to be maintained with new VPP versions.